[fix bug 1240862] Merge two user accounts into one.

akatsoulas opened the pull request:
* Migrate events to the current account.
* Migrate reports to the current account.
* Migrate event attendance to the current account.
* Migrate polls to the current account.
* Migrate user status to the current account.

~~* Delete old user.~~

akatsoulas commented:
@johngian can you please r this?

johngian commented:
Why do you prefer to do this with a database migrations rather than adding eg. a util function in /admin ? I don't really like the idea of using a migration for that. Also the code is going to be the same but more generic, meaning that we can also use this in the future.

johngian commented:
Here is the list of FK and M2M referring to ``User`` not migrated:

* Poll
 * created_by
 * users_vote
* Vote
 * user
* PollComment
 * user
* EventComment
 * user
* UserProfile
 * added_by
 * mentor
* UserStatus
 * user
 * replacement_rep
* NGReport
 * mentor
* NGReportComment
 * user
* ActionItem
 * user
* FeaturedRep
 *  created_by
 *  user
* Bug
 * creator
 * assigned_to
 * cc
 * budget_needinfo

akatsoulas commented:
I don't like the idea either. The reason for choosing to go this way is that merging accounts it's not something that it's supported and won't be in the future. The need for this migration comes from an error and it's one off. Specifically instead of re-activating and old account a new one was created.

Although this is going to be a destructive migration, in terms that the old account is going to be deleted, I removed the `delete()` method. If this migration run smoothly on prod, then we can manually delete the account. 

Regarding your comment about some missing `FK`s, I updated the migration to include the most important ones. A few of the are going to be deleted, due to the cascade delete, but I feel that it's the right way to go. Keeping eg action items from the old account only confusion is going to add. Moreover these kind of data won't affect the historic data.

Btw @johngian, pr updated :)

akatsoulas commented:
cc @nukeador 

Data that we are going to migrate to the new profile:
* Events as creator
* Events attended
* Reports
* Votes
* Polls
* User statuses
* Mentees

Data that will be deleted:
* Action items
* Any kind of comment made
* Needinfo, assigned_to and cc from bugs.

nukeador commented:
Seems good for me, great work guys! :)

johngian commented on remo/profiles/migrations/0060_merge_account.py:
```diff
@@ -0,0 +1,197 @@
+            # Re-assign reports to the current user
+            old_user = old_profile.user
+            current_user = current_profile.user
+            NGReport.objects.filter(user=old_user).update(user=current_user)
```
Why are you using the model directly here and not the frozen ``orm[NGReport]``?

johngian replied:
These also has other occurences.

akatsoulas replied:
I tried this with frozen models and it does not work.

johngian commented:
Code looks OK. r+wc
I also suggest to coordinate with a DBA to get a backup right before we release this on prod.
